rest-api/src/main/scala/za/co/absa/enceladus/rest_api/controllers/v3/RunControllerV3.scala

```diff
@@ -0,0 +1,215 @@
+    getRunForRunIdExpression(datasetName, datasetVersion, runId).map(_.controlMeasure.checkpoints)
+  }
+
+  @PostMapping(Array("/{datasetName}/{datasetVersion}/{runId}/checkpoints"))
+  @ResponseStatus(HttpStatus.CREATED)
+  def addCheckpoint(
+                     @PathVariable datasetName: String,
+                     @PathVariable datasetVersion: Int,
+                     @PathVariable runId: Int,
+                     @RequestBody newCheckpoint: Checkpoint,
+                     request: HttpServletRequest): CompletableFuture[ResponseEntity[String]] = {
+    runService.addCheckpoint(datasetName, datasetVersion, runId, newCheckpoint).map { _ =>
+        val location: URI = ServletUriComponentsBuilder.fromRequest(request)
+          .path("/{cpName}")
+          .buildAndExpand(newCheckpoint.name)
+          .toUri
+        ResponseEntity.created(location).body(s"Checkpoint '${newCheckpoint.name}' added.")
+    }
+  }
+
+  @GetMapping(Array("/{datasetName}/{datasetVersion}/{runId}/checkpoints/{checkpointName}"))
+  @ResponseStatus(HttpStatus.OK)
+  def getRunCheckpointByName(@PathVariable datasetName: String,
+                             @PathVariable datasetVersion: Int,
+                             @PathVariable runId: String,
+                             @PathVariable checkpointName: String): CompletableFuture[Checkpoint] = {
+    for {
+      checkpoints <- getRunForRunIdExpression(datasetName, datasetVersion, runId).map(_.controlMeasure.checkpoints)
+      cpByName = checkpoints.find(_.name == checkpointName).getOrElse(throw NotFoundException())
+    } yield cpByName
+  }
+
+  // todo pagination #2060 ???
+  @GetMapping(Array("/{datasetName}/{datasetVersion}/{runId}/metadata"))
+  @ResponseStatus(HttpStatus.OK)
+  def getRunMetadata(@PathVariable datasetName: String,
+                     @PathVariable datasetVersion: Int,
+                     @PathVariable runId: String): CompletableFuture[ControlMeasureMetadata] = {
+    getRunForRunIdExpression(datasetName, datasetVersion, runId).map(_.controlMeasure.metadata)
+  }
+
+  /**
+   * For run's dataset name, version and runId (either a number of 'latest'), the `forVersionFn` is called.
+   */
+  protected def forRunIdExpression[T](datasetName: String, datasetVersion: Int, runIdStr: String)
+                                     (forVersionFn: (String, Int, Int) => Future[T]): Future[T] = {
+    getRunForRunIdExpression(datasetName, datasetVersion, runIdStr).flatMap { run =>
+      forVersionFn(datasetName, datasetVersion, run.runId)
+    }
+  }
+
+  /**
+   * Retrieves a Run by dataset name, version and runId (either a number of 'latest')
+   *
+   * @param datasetName    dataset name
+   * @param datasetVersion dateset version
+   * @param runIdStr       runId (either a number of 'latest')
+   * @return Run object
+   */
+  protected def getRunForRunIdExpression(datasetName: String, datasetVersion: Int, runIdStr: String): Future[Run] = {
```

Comment:
Why `protected` and not `private`?

Reply:
Should the RunControllerV3 ever get extended, it could get reused in the ancestor class, too.